File: src/logger.ts
import { format } from "node:util";
import { hostname } from "node:os";
import { originalConsole, printLogRecord } from "./console.ts";
import { EventBuilder } from "./event.ts";
import { inferGitBranch, inferGitProject } from "./git.ts";
import { Transport } from "./transport.ts";
import type { LogLevel, LogRecord, LoggerOptions } from "./types.ts";
import { LOG_LEVELS } from "./types.ts";

const HOSTNAME = hostname();
const PID = process.pid;

function resolveLevel(explicit?: LogLevel): LogLevel {
	if (explicit) return explicit;
	const env = process.env.LOG_LEVEL ?? process.env.RELOG_LEVEL;
	if (env && env in LOG_LEVELS) return env as LogLevel;
	return "info";
}

function resolveAuth(explicit?: string): string | undefined {
	return explicit ?? process.env.RELOG_AUTH;
}

function serializeError(err: Error, depth = 0): Record<string, unknown> {
	const out: Record<string, unknown> = {
		error: err.message,
		name: err.name,
		stack: err.stack,
	};
	// Recursively capture the standard `Error.cause` chain (ES2022). Bound
	// recursion at 5 levels so a cycle (rare but possible) can't blow up
	// the serializer.
	if (err.cause !== undefined && depth < 5) {
		out.cause = err.cause instanceof Error ? serializeError(err.cause, depth + 1) : err.cause;
	}
	return out;
}

export class Logger {
	private transport: Transport | null;
	private service: string | undefined;
	private level: LogLevel;
	private consoleEnabled: boolean;
	private boundMeta: Record<string, unknown>;
	private traceId: string | undefined;
	private spanId: string | undefined;
	private project: string | undefined;
	private branch: string | undefined;
	private version: string | undefined;
	private deploymentId: string | undefined;
	private host: string;
	private pid: number;
	private isChild: boolean;
	private sampleRate: number | undefined;
	private slowThresholdMs: number | undefined;

	constructor(options: LoggerOptions = {}, parentTransport?: Transport) {
		this.service = options.service;
		this.level = resolveLevel(options.level);
		this.boundMeta = options.meta ?? {};
		this.traceId = options.traceId;
		this.spanId = options.spanId;
		this.project = options.project ?? inferGitProject();
		this.branch = options.branch ?? inferGitBranch();
		this.version = options.version;
		this.deploymentId = options.deploymentId;
		this.host = HOSTNAME;
		this.pid = PID;
		this.isChild = !!parentTransport;
		this.sampleRate = options.sampleRate;
		this.slowThresholdMs = options.slowThresholdMs;

		this.consoleEnabled = options.console ?? process.env.NODE_ENV !== "production";

		if (parentTransport) {
			this.transport = parentTransport;
		} else if (options.url) {
			this.transport = new Transport({
				url: options.url,
				auth: resolveAuth(options.auth),
				batchSize: options.batchSize,
				flushInterval: options.flushInterval,
				maxBufferSize: options.maxBufferSize,
				onError: options.onError,
			});
		} else {
			this.transport = null;
		}
	}

	child(
		meta: Record<string, unknown> & {
			traceId?: string;
			spanId?: string;
			project?: string;
			branch?: string;
			version?: string;
			deploymentId?: string;
			sampleRate?: number;
			slowThresholdMs?: number;
		},
	): Logger {
		const {
			traceId,
			spanId,
			project,
			branch,
			version,
			deploymentId,
			sampleRate,
			slowThresholdMs,
			...rest
		} = meta;
		const childLogger = new Logger(
			{
				service: this.service,
				level: this.level,
				console: this.consoleEnabled,
				meta: { ...this.boundMeta, ...rest },
				traceId: traceId ?? this.traceId,
				spanId: spanId ?? this.spanId,
				project: project ?? this.project,
				branch: branch ?? this.branch,
				version: version ?? this.version,
				deploymentId: deploymentId ?? this.deploymentId,
				sampleRate: sampleRate ?? this.sampleRate,
				slowThresholdMs: slowThresholdMs ?? this.slowThresholdMs,
			},
			this.transport ?? undefined,
		);
		return childLogger;
	}

	setLevel(level: LogLevel): void {
		this.level = level;
	}

	private log(
		level: LogLevel,
		messageOrError: string | Error,
		meta?: Record<string, unknown>,
	): void {
		if (LOG_LEVELS[level] < LOG_LEVELS[this.level]) return;

		let message: string;
		let finalMeta: Record<string, unknown> | undefined;

		if (messageOrError instanceof Error) {
			message = messageOrError.message;
			finalMeta = {
				...this.boundMeta,
				...serializeError(messageOrError),
				...meta,
			};
		} else {
			message = messageOrError;
			const merged = { ...this.boundMeta, ...meta };
			finalMeta = Object.keys(merged).length > 0 ? merged : undefined;
		}

		const record: LogRecord = {
			timestamp: new Date().toISOString(),
			level,
			message,
			meta: finalMeta,
			service: this.service,
			host: this.host,
			pid: this.pid,
			trace_id: this.traceId,
			span_id: this.spanId,
			project: this.project,
			branch: this.branch,
			version: this.version,
			deployment_id: this.deploymentId,
		};
		// Promote duration_ms from meta to top-level for the DB column
		if (typeof finalMeta?.duration_ms === "number") {
			record.duration_ms = finalMeta.duration_ms;
		}

		if (this.consoleEnabled) {
			printLogRecord(record);
		}

		if (this.transport) {
			this.transport.send(record);
		}
	}

	trace(message: string | Error, meta?: Record<string, unknown>): void {
		this.log("trace", message, meta);
	}

	debug(message: string | Error, meta?: Record<string, unknown>): void {
		this.log("debug", message, meta);
	}

	info(message: string | Error, meta?: Record<string, unknown>): void {
		this.log("info", message, meta);
	}

	warn(message: string | Error, meta?: Record<string, unknown>): void {
		this.log("warn", message, meta);
	}

	error(message: string | Error, meta?: Record<string, unknown>): void {
		this.log("error", message, meta);
	}

	fatal(message: string | Error, meta?: Record<string, unknown>): void {
		this.log("fatal", message, meta);
	}

	event(name: string, meta?: Record<string, unknown>): EventBuilder {
		return new EventBuilder(
			name,
			(level, message, m) => this.log(level, message, m),
			{ ...this.boundMeta, ...meta },
			{ sampleRate: this.sampleRate, slowThresholdMs: this.slowThresholdMs },
		);
	}

	captureConsole(): void {
		const logger = this;
		// Guard against recursion: transport/onError or even user code reachable
		// from log() may eventually call console.* again. Without this, the
		// transport's failure-warning console.warn ends up routed back through
		// the logger and re-buffered indefinitely.
		const wrap = (fn: (msg: string, m?: Record<string, unknown>) => void) => {
			return (...args: unknown[]) => {
				if (Logger._inCapturedConsole) {
					originalConsole.log(...args);
					return;
				}
				Logger._inCapturedConsole = true;
				try {
					fn.call(logger, format(...args), { source: "console" });
				} finally {
					Logger._inCapturedConsole = false;
				}
			};
		};
		console.log = wrap(this.info);
		console.info = wrap(this.info);
		console.warn = wrap(this.warn);
		console.error = wrap(this.error);
		console.debug = wrap(this.debug);
	}

	private static _inCapturedConsole = false;

	restoreConsole(): void {
		console.log = originalConsole.log;
		console.info = originalConsole.info;
		console.warn = originalConsole.warn;
		console.error = originalConsole.error;
		console.debug = originalConsole.debug;
	}

	async flush(): Promise<void> {
		if (this.transport) {
			await this.transport.flush();
		}
	}

	async destroy(): Promise<void> {
		if (this.transport && !this.isChild) {
			await this.transport.flush();
			this.transport.destroy();
		}
	}
}

export function createLogger(options: LoggerOptions = {}): Logger {
	return new Logger(options);
}

File: python/src/relog/_logger.py
from __future__ import annotations

import os
import socket
import traceback
from datetime import datetime, timezone
from typing import Any

from ._console import print_log_record
from ._event import EventBuilder
from ._git import infer_git_branch, infer_git_project
from ._transport import Transport
from ._types import LOG_LEVELS, LogLevel, LogRecord, LoggerOptions, SamplingOptions

_HOSTNAME = socket.gethostname()
_PID = os.getpid()


def _resolve_level(explicit: LogLevel | None) -> LogLevel:
    if explicit:
        return explicit
    env = os.environ.get("LOG_LEVEL") or os.environ.get("RELOG_LEVEL")
    if env and env in LOG_LEVELS:
        return env  # type: ignore[return-value]
    return "info"


def _resolve_auth(explicit: str | None) -> str | None:
    return explicit or os.environ.get("RELOG_AUTH")


def _serialize_error(err: BaseException) -> dict[str, Any]:
    return {
        "error": str(err),
        "name": type(err).__name__,
        "stack": "".join(
            traceback.format_exception(type(err), err, err.__traceback__)
        ),
    }


class Logger:
    def __init__(
        self,
        options: LoggerOptions | None = None,
        _parent_transport: Transport | None = None,
    ) -> None:
        opts = options or LoggerOptions()

        self._service = opts.service
        self._level = _resolve_level(opts.level)
        self._console_enabled = opts.console if opts.console is not None else True
        self._bound_meta: dict[str, Any] = dict(opts.meta) if opts.meta else {}
        self._trace_id = opts.trace_id
        self._span_id = opts.span_id
        self._project = opts.project or infer_git_project()
        self._branch = opts.branch or infer_git_branch()
        self._version = opts.version
        self._deployment_id = opts.deployment_id
        self._host = _HOSTNAME
        self._pid = _PID
        self._is_child = _parent_transport is not None
        self._sample_rate = opts.sample_rate
        self._slow_threshold_ms = opts.slow_threshold_ms

        if _parent_transport:
            self._transport: Transport | None = _parent_transport
        elif opts.url:
            self._transport = Transport(
                url=opts.url,
                auth=_resolve_auth(opts.auth),
                batch_size=opts.batch_size or 50,
                flush_interval=opts.flush_interval or 5.0,
                max_buffer_size=opts.max_buffer_size or 10_000,
                on_error=opts.on_error,
            )
        else:
            self._transport = None

    def child(
        self,
        *,
        trace_id: str | None = None,
        span_id: str | None = None,
        project: str | None = None,
        branch: str | None = None,
        version: str | None = None,
        deployment_id: str | None = None,
        sample_rate: float | None = None,
        slow_threshold_ms: float | None = None,
        meta: dict[str, Any] | None = None,
    ) -> Logger:
        merged_meta = {**self._bound_meta, **(meta or {})}
        child_opts = LoggerOptions(
            service=self._service,
            level=self._level,
            console=self._console_enabled,
            meta=merged_meta,
            trace_id=trace_id or self._trace_id,
            span_id=span_id or self._span_id,
            project=project or self._project,
            branch=branch or self._branch,
            version=version or self._version,
            deployment_id=deployment_id or self._deployment_id,
            sample_rate=sample_rate if sample_rate is not None else self._sample_rate,
            slow_threshold_ms=slow_threshold_ms
            if slow_threshold_ms is not None
            else self._slow_threshold_ms,
        )
        return Logger(child_opts, self._transport or None)

    def set_level(self, level: LogLevel) -> None:
        self._level = level

    def _log(
        self,
        level: LogLevel,
        message_or_error: str | BaseException,
        meta: dict[str, Any] | None = None,
    ) -> None:
        if LOG_LEVELS[level] < LOG_LEVELS[self._level]:
            return

        if isinstance(message_or_error, BaseException):
            message = str(message_or_error)
            final_meta = {
                **self._bound_meta,
                **_serialize_error(message_or_error),
                **(meta or {}),
            }
        else:
            message = message_or_error
            merged = {**self._bound_meta, **(meta or {})}
            final_meta = merged if merged else None  # type: ignore[assignment]

        record = LogRecord(
            timestamp=datetime.now(timezone.utc).isoformat(),
            level=level,
            message=message,
            meta=final_meta,
            service=self._service,
            host=self._host,
            pid=self._pid,
            trace_id=self._trace_id,
            span_id=self._span_id,
            project=self._project,
            branch=self._branch,
            version=self._version,
            deployment_id=self._deployment_id,
        )

        if self._console_enabled:
            print_log_record(record)

        if self._transport:
            self._transport.send(record)

    def trace(
        self, message: str | BaseException, meta: dict[str, Any] | None = None
    ) -> None:
        self._log("trace", message, meta)

    def debug(
        self, message: str | BaseException, meta: dict[str, Any] | None = None
    ) -> None:
        self._log("debug", message, meta)

    def info(
        self, message: str | BaseException, meta: dict[str, Any] | None = None
    ) -> None:
        self._log("info", message, meta)

    def warn(
        self, message: str | BaseException, meta: dict[str, Any] | None = None
    ) -> None:
        self._log("warn", message, meta)

    def error(
        self, message: str | BaseException, meta: dict[str, Any] | None = None
    ) -> None:
        self._log("error", message, meta)

    def fatal(
        self, message: str | BaseException, meta: dict[str, Any] | None = None
    ) -> None:
        self._log("fatal", message, meta)

    def event(
        self, name: str, meta: dict[str, Any] | None = None
    ) -> EventBuilder:
        sampling = SamplingOptions(
            sample_rate=self._sample_rate,
            slow_threshold_ms=self._slow_threshold_ms,
        )
        return EventBuilder(
            name,
            lambda level, msg, m: self._log(level, msg, m),
            {**self._bound_meta, **(meta or {})},
            sampling,
        )

    def flush(self) -> None:
        if self._transport:
            self._transport.flush()

    def destroy(self) -> None:
        if self._transport and not self._is_child:
            self._transport.flush()
            self._transport.destroy()


def create_logger(
    url: str | None = None,
    *,
    service: str | None = None,
    auth: str | None = None,
    level: LogLevel | None = None,
    batch_size: int | None = None,
    flush_interval: float | None = None,
    max_buffer_size: int | None = None,
    console: bool | None = None,
    meta: dict[str, Any] | None = None,
    trace_id: str | None = None,
    span_id: str | None = None,
    project: str | None = None,
    branch: str | None = None,
    version: str | None = None,
    deployment_id: str | None = None,
    sample_rate: float | None = None,
    slow_threshold_ms: float | None = None,
    on_error: Any | None = None,
) -> Logger:
    opts = LoggerOptions(
        url=url,
        service=service,
        auth=auth,
        level=level,
        batch_size=batch_size,
        flush_interval=flush_interval,
        max_buffer_size=max_buffer_size,
        console=console,
        meta=meta,
        trace_id=trace_id,
        span_id=span_id,
        project=project,
        branch=branch,
        version=version,
        deployment_id=deployment_id,
        sample_rate=sample_rate,
        slow_threshold_ms=slow_threshold_ms,
        on_error=on_error,
    )
    return Logger(opts)
